### DISCUSSION TITLE
`open "my_ls.db" | db query "select * from main"` shouldn't have been `db open` instead?

### DISCUSSION THREAD

**Discussion by shemeshg** [nushell/nushell#12446] [General]
```nu
> let a = open "my_ls.db"
> $a | describe                             
SQLiteDatabase
- $a | db query "select * from main"
```

1. The given Nushell code snippet appears to open a database file named “my_ls.db” and execute a query on it. The output seems to be related to a SQLiteDatabase. However, why and how does the db command access the database without managing (owning) the connection?

    isn't the connection managed by the `open`?

2. should custom structs like `SQLiteDatabase`be restricted to streaming only to and from the object that created them (have hidden uuid, and validate against that)?

* could restricting custom structs to stream only to the object that created them  enhance encapsulation and prevent unintended side effects?

Maybe something like
```nu
> let a =  db open "my_ls.db"
> "select * from main"|db query $a
> db close $a
> $a|db describe
```

`$a|describe`, could print "Custom struct owned by db".
`$a` `<enter>` , could raise exception of "Custom struct owned by db, only db can view it's content" and not implicit load of all db.

**Comment by devyn** [comment]
Hi, we don't have a need for an explicit `db close` command because management of objects is tied to their scope. If you use something in a function or command, unless you return it, it will be freed or closed automatically when you leave that scope.

Changing the behavior of custom values on print is totally possible, we just have it print all of the tables for convenience. If you make your own custom values in a plugin you can change them to do whatever you like.

**Reply by shemeshg** [reply]
regards the no need for `db close`

I have tried in the example plugin in the contribution documentation to

```
impl Drop for LenPlugin {
    fn drop(&mut self) {
        eprintln!("Out of scope");
    }
}
```

and from nu
```
 do {                     08/04/24 11:41:51
  let a = len "asd"
 }
```

This did not trigger the cleanup. (though **a** is out of scope)
I think the plugin is not aware of the scope, and remains in memory (as expected and should) , as long as required by the `register` in `nu`

It is not a bad thing, because it helps keeping the connection alive, hwever not sure it can implicitly closing it

**Reply by devyn** [reply]
The `Drop` impl has to be on a custom value, and that custom value has to implement

```rust
fn notify_plugin_on_drop(&self) -> bool {
    true
}
```

Implementing it on the plugin will not work until the plugin stops.

You can look at the tests in the custom value plugin in the repo - specifically, for DropCheck. We test them automatically, so they should work.

The out of scope drop feature is a Rust feature that we also reuse in Nushell.

**Reply by shemeshg** [reply]
Thanks, 

<https://github.com/nushell/nushell/blob/main/crates/nu_plugin_custom_values/src/drop_check.rs>
helped

I've managed to create custom values, yet it did not triggered the 
```rust
    pub(crate) fn notify(&self) {
        eprintln!("DropCheckValue was dropped: {}", self.msg);
    }
```

and I had to remark `fn as_mut_any(&mut self)` because of
```text
     /     fn as_mut_any(&mut self) -> &mut dyn std::any::Any {
50 | |         self
51 | |     }
   | |_____^ not a member of trait `CustomValue`
```

However I'm sure it is something I'm missing out, and I will retry it later.

```rust
use nu_plugin::{serve_plugin, EvaluatedCall, JsonSerializer};
use nu_plugin::{EngineInterface, Plugin, PluginCommand, SimplePluginCommand};
use nu_protocol::{CustomValue, LabeledError, ShellError, Signature, SyntaxShape, Type, Value,Span,record};
use serde::{Deserialize, Serialize};


#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct DropCheckValue {
    pub(crate) msg: String,
}

impl DropCheckValue {
    pub(crate) fn new(msg: String) -> DropCheckValue {
        DropCheckValue { msg }
    }

    pub(crate) fn into_value(self, span: Span) -> Value {
        Value::custom(Box::new(self), span)
    }

    pub(crate) fn notify(&self) {
        eprintln!("DropCheckValue was dropped: {}", self.msg);
    }
}

#[typetag::serde]
impl CustomValue for DropCheckValue {
    fn clone_value(&self, span: Span) -> Value {
        self.clone().into_value(span)
    }

    fn type_name(&self) -> String {
        "DropCheckValue".into()
    }

    fn to_base_value(&self, span: Span) -> Result<Value, ShellError> {
        Ok(Value::record(
            record! {
                "msg" => Value::string(&self.msg, span)
            },
            span,
        ))
    }

    fn as_any(&self) -> &dyn std::any::Any {
        self
    }

    /* 
     /     fn as_mut_any(&mut self) -> &mut dyn std::any::Any {
50 | |         self
51 | |     }
   | |_____^ not a member of trait `CustomValue`
   */
    /*
    fn as_mut_any(&mut self) -> &mut dyn std::any::Any {
        self
    }
    */

    fn notify_plugin_on_drop(&self) -> bool {
        // This is what causes Nushell to let us know when the value is dropped
        true
    }
}
```

**Reply by devyn** [reply]
You may have been trying to compile this for the wrong version. `as_mut_any` shipped with 0.92.1

The drop does work, just try `custom-value drop-check hello` on its own and you should see the message. You can also try letting it to a variable inside a scope like a command or `do { ... }` and the same will happen.

On top of supporting it for plugins, this is just a basic feature of how destructors in Rust work, we don't have to manually clean up resources.
